### PR TITLE
Improve the metrics user experience 

### PR DESCRIPTION
**Pull request by danielmarbach**
by moving relevant parts to the feature and make the feature disable itself on send-only endpoints instead of throwing

outcome of feedback in https://github.com/Particular/NServiceBus.Metrics/commit/e6932a2e8e3b117a7633cfa194ad79844274a46a#r138177362

### CODE REVIEW

File: src/NServiceBus.Metrics/MetricsConfigurationExtensions.cs [hunk 1]
```diff
@@ -19,13 +19,8 @@ public static MetricsOptions EnableMetrics(this EndpointConfiguration endpointCo
...
 
             var settings = endpointConfiguration.GetSettings();
-            var options = settings.GetOrCreate<MetricsOptions>();
-            settings.Set(typeof(MetricsFeature).FullName, FeatureState.Enabled);
-
-            endpointConfiguration.Recoverability().Immediate(c => c.OnMessageBeingRetried((m, ct) => options.Immediate(m, ct)));
-            endpointConfiguration.Recoverability().Delayed(c => c.OnMessageBeingRetried((m, ct) => options.Delayed(m, ct)));
-
-            return options;
+            settings.EnableFeatureByDefault<MetricsFeature>();
```
Comment (danielmarbach) [inline src/NServiceBus.Metrics/MetricsConfigurationExtensions.cs:22]: Is there a subtle difference to settings.Set(typeof(MetricsFeature).FullName, FeatureState.Enabled); ?

Comment (andreasohlund) [reply src/NServiceBus.Metrics/MetricsConfigurationExtensions.cs:22]: The way I remember it:

`settings.EnableFeatureByDefault<MetricsFeature>();` => on by default but if a user has disabled it or one of the prereqs are not true it won't be enabled

`settings.Set(typeof(MetricsFeature).FullName, FeatureState.Enabled);` => force it enabled no matter what

**Comment by bbrandt** [comment]
Wow!  I am just so excited about this change.  You're my hero, @danielmarbach!

File: src/NServiceBus.Metrics/MetricsConfigurationExtensions.cs (hunk 1 above)
Comment (danielmarbach) [reply src/NServiceBus.Metrics/MetricsConfigurationExtensions.cs:22]: @andreasohlund under that light I think `settings.EnableFeatureByDefault<MetricsFeature>();` is the better choice, no?

Comment (andreasohlund) [reply src/NServiceBus.Metrics/MetricsConfigurationExtensions.cs:22]: Agree

**Comment by bording** [comment]
@danielmarbach Looking through these changes, I'm not yet convinced this is actually a better experience overall. Yes, you've removed the exception being thrown, but now you can create a send-only endpoint that has a call to `EnableMetrics` that just silently gets ignored, leaving someone to think that metrics are enabled, when they actually aren't.

I'm not a fan of having APIs that can't be combined, but then provide no feedback that one of them is a no-op like that.

**Comment by danielmarbach** [comment]
> that has a call to EnableMetrics that just silently gets ignored

FYI Features that do not meet the prerequisites are logged with a reason. 

I'm good either way, to be honest. Looking at our history of designing those APIs arguments can go both ways. For example we have certain types of APIs that are related to features that are enabled by default and disable themselves when the requirements are not met. For example, sagas is such a feature that "silently" disables itself when there are no saga types in the endpoint.

Then we have other APIs like `EnableOutbox` or the audit feature that activates a feature that then gets disabled "silently" when the requirements are not met.

Finally, we have this feature and, for example also the gateway that do a check for send-only and then throw instead of disabling themselves via prerequisites. 

This shows that there is no clear winner approach and mileage may vary. 

From a general standpoint one could argue having the feature disable itself as long as the lib doesn't provide metrics for send-only endpoints is a good approach because it makes it possible to add send-only support which then would enable the feature or parts of the feature for the metrics relevant for send-only. I realize this is a bit of a straw man argument though here because Metrics is probably not likely to evolve a lot given NSB supports now open telemetry. 

In absence of a clear winner approach I guess we have to find a common principle to agree upon. 

@bbrandt feel free to share your thoughts.

**Comment by bbrandt** [comment]
Yes.  I had been considering @bording's concern as well.  I agree that if there were no notification/logging that the feature was being disabled it would be violating principal of least surprise.  However, we were trying to internally publish a small NuGet package for our NSB best-practices, with our [PromethusFeature](https://github.com/Particular/docs.particular.net/blob/b589d66bcb557457a7fcb8af335fcd04309975bc/samples/logging/prometheus-grafana/Metrics_4/Endpoint/PrometheusFeature.cs), which we adapted to automatically disable is a service adds the SendOnly feature.  We thought, "All services should enable metrics", so we put that in our standard NSB config as well.  We were then surprised that Metrics threw an exception for our API services that try to add the SendOnly feature.  I am not fully convinced the approach we've chosen of having a shared NuGet package for NSB config is the best approach, but it is the one we chose as a team.

There is certainly a tradeoff here of simplicity of configuration vs fail fast that we are trying to come to a good balance on.  But since we do have the output to let users know the metrics will not be able to be enabled because it is not compatible with the SendOnly feature, I feel like your PR is a positive change.  

As you mentioned, it is only helpful for those of us still limping along on the older metrics lib, not yet migrated to OpenTelemetry.  We are currently using the Classic App Insights SDK with a bit of code written directly against the AI-sdk, unfortunately. We still have about 6-9 months of other work prioritized above cleaning this up and migrating to Otel.  Since we are still pretty far out from this, I do greatly appreciate you taking the time to make this change.

**Comment by bording** [comment]
@bbrandt If you have centralized configuration, you must have something that determines if an endpoint should be send-only or not as part of it. The way I would solve this would be to use that setting to skip calling `EnableMetrics` when an endpoint is going to be send-only. 

This is something that is completely in your control with the existing APIs, so it's not clear to me why this change is needed at all.

I've also consulted with some other collogues offline about this, and I'm increasingly leaning towards this being a change that we don't want to make because of it silently hiding something that should not be hidden.

File: src/NServiceBus.Metrics/MetricsFeature.cs [hunk 2]
```diff
@@ -12,11 +14,26 @@
...
+        /// 
+        /// </summary>
+        public MetricsFeature()
+        {
+            Prerequisite(c => !c.Settings.Get<bool>("Endpoint.SendOnly"), "Processing metrics are not supported on send-only endpoints");
+            Defaults(settings =>
+            {
+                var bindingFlags = BindingFlags.NonPublic | BindingFlags.Instance;
+                // Unfortunately the constructor is internal, so we have to use reflection to create an instance of RecoverabilitySettings
+                var recoverability = (RecoverabilitySettings)Activator.CreateInstance(typeof(RecoverabilitySettings), bindingFlags, null, [settings], null);
```
Comment (bbrandt) [inline src/NServiceBus.Metrics/MetricsFeature.cs:27]: Does this newly created `RecoverabilitySettings` have to somehow be connected to the `EndpointConfiguration` for it to take effect?

**Comment by bbrandt** [comment]
Yes, @bording.  You are absolutely right.  We have all the control we need right now, but this started as the question I pasted below about why there was code to configure Recoverability in the `EnableMetrics()` extension method rather than this being in feature setup.  

In our [`ConfigureNServiceBus()`](https://gist.github.com/bbrandt/b3a8b80835776a63e664ea0e9b5ff698#file-hostingextensions-cs-L35) we were asking all microservice teams to use, we had called `configuration.EnableMetrics()`, thinking surely we want every single endpoint to be producing metrics.  But as soon as someone added [`IsSendOnly()`](https://gist.github.com/bbrandt/b3a8b80835776a63e664ea0e9b5ff698#file-servicebusendpointextensions-cs-L13) to the EndpointConfiguration, they were greeted with an exception.  This exception being thrown, while inconvenient in the moment, allowed us to learn that Metrics is not compatible with SendOnly endpoints.  We likely would have never noticed a log message that said metrics would not be enabled since SendOnly is being used. 

I really like the cleanup done in the PR and that the configuration code in question moved out of `EnableMetrics()`.  

@bording Would just restoring the `context.ThrowIfSendonly();` in [Setup()](https://github.com/Particular/NServiceBus.Metrics/pull/530/files#diff-b08485bfeaf5f920a7031144617389fb608ff27479880fad48909c9c128ec0f7R35) (and likely needing to remove the [Prerequisite](https://github.com/Particular/NServiceBus.Metrics/pull/530/files#diff-b08485bfeaf5f920a7031144617389fb608ff27479880fad48909c9c128ec0f7R35)) be enough to get this to move forward?

--
https://github.com/Particular/NServiceBus.Metrics/commit/e6932a2e8e3b117a7633cfa194ad79844274a46a#r138180961
> Why is this not done in [Feature setup](https://docs.particular.net/nservicebus/pipeline/features#feature-setup) or a [Feature startup tasks](https://docs.particular.net/nservicebus/pipeline/features#feature-startup-tasks)?

> On Thu, Feb 1, 2024, 2:38 PM Daniel Marbach ***@***.***> wrote:
 Don't recall specifics but I think it is due to the feature setup APIs not
 being the official extension point to access the recoverability APIs. Are
 you currently facing issues?

https://github.com/Particular/NServiceBus.Metrics/commit/e6932a2e8e3b117a7633cfa194ad79844274a46a#r138177362
> No real blocking issue.  We were just looking for a way to have metrics
automatically disabled when send-only gets enabled.  We manage many
microservice teams and were trying to simplify our standard NSB setup code.
>
> We have a ConfigureNServicebus() extension method to add our standard
behaviors and thought we should always enable metrics, then someone needed
to add SendOnly() and got an exception.
> 
> What I'm thinking now is we should have a separate ConfigureNServicebus()
and ConfigureSendOnlyNServicebus() to encapsulate our best practices for
these 2 very different types of config.
[…](https://github.com/Particular/NServiceBus.Metrics/commit/e6932a2e8e3b117a7633cfa194ad79844274a46a#)

**Comment by danielmarbach** [comment]
@bbrandt Brandon and I had a discussion today and we mutually agreed that the cleanup part is valuable and we will pick some of it into a dedicated PR. We went through the examples of different features behaving differently and concluded the basic principle of a feature throwing under the circumstances of two explicit configuration API calls being inherently incompatible definitely makes more sense rather than silently disabling itself. Given that this API was already designed under that premise and you have a way to discover that the endpoint is configured to be send-only we are leaving the exception in place.

I'll be off a few days so it might take a while until this gets picked up again unless @bording finds some time pushing this forward.

Thanks for all your feedback on this conversation.

File: src/NServiceBus.Metrics/MetricsFeature.cs (hunk 2 above)
Comment (danielmarbach) [reply src/NServiceBus.Metrics/MetricsFeature.cs:27]: That is done by passing the settingsHolder bag into the constructor.

**Comment by bbrandt** [comment]
@danielmarbach I am incredibly happy with the direction you've taken, to fail fast when there is a configuration that is not supported.  Thank you for all your work on this as well as taking the time to consult your team.  Software development is most often an individual creative endeavor, but nothing beats having an excellent group of peers to be able to check ideas against.

**Comment by danielmarbach** [comment]
@bording @bbrandt I had a bit of time during traveling and raised https://github.com/Particular/NServiceBus.Metrics/pull/534

**Pull request closed without merging**